### ISSUE TITLE
Unable to Reproduce Hyperparameter Tuning Fitness Plot from HPT Guide

### ISSUE THREAD

**Issue by Aditya-365** [ultralytics/ultralytics#24361]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hello, I am trying to reproduce the fitness plot shown in the Ultralytics hyperparameter tuning guide but have been unable to achieve similar results.

---

## What I am trying to replicate
The fitness vs iteration plot shown in the HPT documentation, where fitness steadily improves and plateaus over time. The graph is from: https://docs.ultralytics.com/guides/hyperparameter-tuning/#tune_fitnesspng

<img width="491" height="299" alt="Image" src="https://github.com/user-attachments/assets/f00060f3-42bf-456b-a334-ffd73ef786f5" />

---

### Setup
- Model: `yolo26n.pt`
- Dataset(s):
  - coco8
  - coco8 + coco8-grayscale
- Hyperparameters tuned:
  - lr0, momentum, weight_decay, dfl, cls, box
- Epochs: 30
- Iterations: 5 (also tested with higher values)

### Observations
- Fitness curves are noisy and inconsistent  
- No clear upward trend or plateau  
- Very few data points compared to the guide  

---

## Questions

1. **How can I reproduce the fitness plot shown in the documentation?**  
   Are there specific settings (iterations, population size, dataset scale) required?

2. **Why does each iteration in the guide appear to produce multiple data points?**  
   Is the tuner evaluating multiple candidates per iteration (e.g., population-based methods)?

3. **What contributes to the smooth, steadily increasing fitness curve?**  
   Is this due to:
   - large number of iterations (e.g., 500+)  
   - smoothing/aggregation of results  
   - specific dataset or training configuration  

---

## Environment
- OS: Windows 11  
- Python: 3.11   
- GPU: Nvidia RTX 5090  

---

## Code used
~~~python
from ultralytics import YOLO

model = YOLO("yolo26n.pt")
model.tune(
    data="coco8.yaml",
    epochs=30,
    iterations=5,
    space={
        "lr0": (1e-5, 1e-1),
        "momentum": (0.6, 0.98),
        "weight_decay": (0.0, 0.001),
        "dfl": (0.5, 2.0),
        "cls": (0.2, 4.0),
        "box": (0.02, 0.2)
    }
)
~~~
---

## Results
Run 1: COCO8:

<img width="2000" height="1200" alt="Image" src="https://github.com/user-attachments/assets/73911ae6-2a58-4cea-aa6c-9b2c1b20181b" />

Run 2: COCO8 + COCO8-grayscale 

<img width="2000" height="1200" alt="Image" src="https://github.com/user-attachments/assets/d1f01499-4aa2-41db-ab21-66052a1a7920" />

Run 3: COCO8 (Extended iterations) - [50 epcohs and 300 iterations]

<img width="894" height="531" alt="Image" src="https://github.com/user-attachments/assets/293c7d7e-3222-438f-8878-7d96dd96f050" />

### Additional

I understand that my iteration count is relatively small, but I would like to understand the key factors required to reproduce the trend shown in the guide.

Thank you for your time.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Aditya-365, thanks for the detailed report and for your interest in Ultralytics 🚀 This is an automated response to help get things moving, and an Ultralytics engineer will also assist you soon.

We recommend a quick visit to the [Docs](https://docs.ultralytics.com/) for new users, where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many common questions may already be answered.

Since this appears to be a hyperparameter tuning/custom training ❓ question, please provide as much information as possible, including:
- Full `ultralytics` package version
- Full console logs from the tuning run
- Any saved tuning artifacts/results directory contents
- Whether the same behavior occurs after upgrading and rerunning in a clean environment
- If you believe this may be a 🐛 bug or documentation mismatch, please also share a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us investigate

Please also verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) and, if possible, include any additional screenshots or logs that show how the plot evolves across runs 📈

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
You're not missing a hidden population setting here: the built-in tuner evaluates one mutated hyperparameter set per `iteration` and appends one record to `tune_results.ndjson`, so there is no population-size parameter in the internal tuner; in `tune_fitness.png` the extra-looking points are the per-dataset series, and the smooth rising curve is a Gaussian-smoothed mean from `plot_tune_results()` rather than raw monotonic fitness. For multi-dataset runs the top-level `fitness` is just the average across datasets, and with toy data like `coco8` plus short 30-epoch trainings, variance will dominate, so the docs figure is best treated as a long-run illustrative example rather than something you should expect to reproduce exactly on small runs; see the [tuner reference](https://docs.ultralytics.com/reference/engine/tuner/), [plotting reference](https://docs.ultralytics.com/reference/utils/plotting/), and the note in the [hyperparameter tuning guide](https://docs.ultralytics.com/guides/hyperparameter-tuning/).

**Comment by raimbekovm** [comment]
hi @Aditya-365, dug into the tuner code and ran two short reproductions on coco8 to verify. a few clarifications:

**per iteration:** the tuner generates **one** mutated hyperparameter set and writes **one** record to `tune_results.ndjson`. there is no `population_size`, only `iterations`. iteration 1 uses defaults, evolution kicks in from iteration 2.

**how `tune_fitness.png` is built:** each dataset contributes one dot per iteration, plus a gaussian-smoothed mean curve (sigma=3). single-dataset run = 1 dot per iteration, multi-dataset = N dots per iteration. so the "many points" you see in the docs figure are not multiple candidates per iteration, they are simply many iterations. the example image comes from a long run (the `best_hyperparameters.yaml` snippet just below it shows `558/900 iterations complete`).

**why your runs look noisy:**
1. 5 iterations is too few for the smoothed mean to show a trend.
2. coco8 has 8 images, so per-trial fitness variance dominates any signal between hyperparameter sets. your run 3 (300 iter, 50 ep) confirms this: more dots, no trend, because the dataset is too small.
3. the mutation sigma is calibrated to decay over the first 300 iterations, so the algorithm is designed for hundreds of iterations.

**to get a docs-like trend:** tune on the dataset you actually plan to train on (or at least coco128-sized), plan for 300+ iterations, and expect the smooth curve to start showing shape only after the first 50-100.
